base/utf16.jl

```diff
@@ -20,92 +45,248 @@ function endof(s::UTF16String)
+            # output surrogate pair
+            buf[out += 1] = UInt16(0xd7c0 + (ch >>> 10))
+            buf[out += 1] = UInt16(0xdc00 + (ch & 0x3ff))
         end
     end
-    push!(buf, 0) # NULL termination
+    @inbounds buf[out + 1] = 0 # NULL termination
     UTF16String(buf)
 end
 
-utf16(x) = convert(UTF16String, x)
-convert(::Type{UTF16String}, s::UTF16String) = s
-convert(::Type{UTF16String}, s::AbstractString) = encode16(s)
-convert(::Type{Array{UInt16,1}}, s::UTF16String) = s.data
-convert(::Type{Array{UInt16}}, s::UTF16String) = s.data
+"
+Converts a UTF8String to a UTF16String
 
-# TODO: optimize this
-convert(::Type{UTF8String}, s::UTF16String) =
-    sprint(length(s.data)-1, io->for c in s; write(io,c::Char); end)
+### Input Arguments:
+*   ::Type{UTF16String}
+*   str::UTF8String
 
-sizeof(s::UTF16String) = sizeof(s.data) - sizeof(UInt16)
-unsafe_convert{T<:Union(Int16,UInt16)}(::Type{Ptr{T}}, s::UTF16String) =
-    convert(Ptr{T}, pointer(s))
+### Returns:
+*   UTF16String
 
-function isvalid(::Type{UTF16String}, data::AbstractArray{UInt16})
-    i = 1
-    n = length(data) # this may include NULL termination; that's okay
-    while i < n # check for unpaired surrogates
-        if utf16_is_lead(data[i]) && utf16_is_trail(data[i+1])
-            i += 2
-        elseif utf16_is_surrogate(data[i])
-            return false
+### Throws:
+*   UnicodeError
+"
+function convert(::Type{UTF16String}, str::UTF8String)
+    dat = str.data
+    # handle zero length string quickly
+    sizeof(dat) == 0 && return empty_utf16
+    # Check that is correct UTF-8 encoding and get number of words needed
+    len, flags, num4byte = check_string(dat)
+    len += num4byte
+    buf = Vector{UInt16}(len+1)
+    @inbounds buf[len+1] = 0
+    # Optimize case where no characters > 0x7f
+    flags == 0 && @inbounds return UTF16String(copy!(buf, dat))
+    out = 0
+    pos = 0
+    @inbounds while out < len
+        ch::UInt32 = dat[pos += 1]
+        # Handle ASCII characters
+        if ch <= 0x7f
+            buf[out += 1] = ch
+        # Handle range 0x80-0x7ff
+        elseif ch < 0xe0
+            buf[out += 1] = ((ch & 0x1f) << 6) | (dat[pos += 1] & 0x3f)
+        # Handle range 0x800-0xffff
+        elseif ch < 0xf0
+            pos += 2
+            buf[out += 1] = get_utf8_3byte(dat, pos, ch)
+        # Handle range 0x10000-0x10ffff
         else
-            i += 1
+            pos += 3
+            ch = get_utf8_4byte(dat, pos, ch)
+            # output surrogate pair
+            buf[out += 1] = UInt16(0xd7c0 + (ch >>> 10))
+            buf[out += 1] = UInt16(0xdc00 + (ch & 0x3ff))
         end
     end
-    return i > n || !utf16_is_surrogate(data[i])
+    UTF16String(buf)
 end
 
+"
+Converts a UTF-16 encoded vector of UInt16 to a UTF8String
+
+### Input Arguments:
+*   ::Type{UTF8String}
+*   dat::Vector{UInt16}
+
+### Returns:
+*   UTF8String
+
+### Throws:
+*   UnicodeError
+"
+function convert(::Type{UTF8String}, dat::Vector{UInt16})
```

Comment:
i don't see this method in the current list of methods: 
```julia
julia> convert(UTF8String,UInt16[])
ERROR: MethodError: `convert` has no method matching convert(::Type{UTF8String}, ::Array{UInt16,1})
```
therefore it should getting added amidst this PR.

But, from what I can tell, none of the existing methods accept a data in one format and output it in a different format. they are all either String -> String conversions (no interpretation) or Byte -> String interpretation (no conversion).
```julia
julia> convert(UTF8String,\t
convert(::Type{UTF8String}, s::UTF8String) at utf8.jl:215
convert(::Type{UTF8String}, s::ASCIIString) at utf8.jl:216
convert(::Type{UTF8String}, a::Array{UInt8,1}) at utf8.jl:217
convert(::Type{UTF8String}, a::Array{UInt8,1}, invalids_as::AbstractString) at utf8.jl:219
convert(::Type{UTF8String}, s::UTF16String) at utf16.jl:81
convert(::Type{UTF8String}, s::AbstractString) at utf8.jl:239
convert{T<:Union(UTF8String,ASCIIString)}(::Type{T<:Union(UTF8String,ASCIIString)}, data::AbstractArray{Char,1}) at utf32.jl:35
convert{T<:AbstractString}(::Type{T<:AbstractString}, v::AbstractArray{Char,1}) at utf32.jl:31
convert{S<:AbstractString}(::Type{S<:AbstractString}, g::Base.UTF8proc.GraphemeIterator{S<:AbstractString}) at utf8proc.jl:232
convert{T}(::Type{T}, x::T) at essentials.jl:57
julia> convert(UTF8String,
```
so i don't see the value in adding this special case at all. At a minimum, this would be inconsistent. But, it also seems incorrect to me as well, since this particular method is trying to do both interpretation (of UInt16 bytes as a UTF16 native-endian string) and conversion (to UTF8), and then ends up not doing either in a particularly well-defined manner.

> Also, why is having something maximally generic more important than having the functions performant enough to be useful? Other reviewers have agreed that the performance is more important than having the code "maximally generic" (back during the #11004 discussions)

there's a big difference between writing an optimized kernel to hand-optimize certain functions vs. duplicating code in an attempt to subvert the abstraction layer. It is sometimes acceptable to define a hand-optimized method (like `convert(::UTF8String, ::UTF16String)`, rather than always falling back on `convert(::UTF8String, ::AbstractString)`), where the behavior is the same, but optimized in ways that would have been too costly for a compiler to determine.

(breadcrumb: mentioned this issue previously in https://github.com/JuliaLang/julia/pull/11004#discussion_r31568679, but perhaps not quite as clearly)